src/main/java/com/zendesk/maxwell/schema/SchemaStore.java

```diff
@@ -229,8 +266,79 @@ public static SchemaStore restore(Connection connection, MaxwellContext context)
 		return s;
 	}
 
-	private void restoreFrom(Connection conn, BinlogPosition targetPosition) throws SQLException, InvalidSchemaError {
-		PreparedStatement p;
+	private List<ResolvedSchemaChange> parseDeltas(String json) throws IOException {
+		if ( json == null )
+			return null;
+
+		return mapper.readerFor(listOfResolvedSchemaChangeType).readValue(json.getBytes());
+	}
+
+	private HashMap<Long, HashMap<String, Object>> buildSchemaMap(Connection conn) throws SQLException {
+		HashMap<Long, HashMap<String, Object>> schemas = new HashMap<>();
+
+		PreparedStatement p = conn.prepareStatement("SELECT * from `schemas` where server_id = ?");
+		p.setLong(1, this.serverID);
+		ResultSet rs = p.executeQuery();
+
+		ResultSetMetaData md = rs.getMetaData();
+		while ( rs.next() ) {
+			HashMap<String, Object> row = new HashMap<>();
+			for ( int i = 1; i <= md.getColumnCount(); i++ )
+				row.put(md.getColumnName(i), rs.getObject(i));
+			schemas.put(rs.getLong("id"), row);
+		}
+		rs.close();
+		return schemas;
+	}
+
+	private LinkedList<Long> buildSchemaChain(HashMap<Long, HashMap<String, Object>> schemas, Long schema_id) {
+		LinkedList<Long> schemaChain = new LinkedList<>();
+
+		while ( schema_id != null ) {
+			if ( !schemas.containsKey(schema_id) )
+				throw new RuntimeException("Couldn't find chained schema: " + schema_id);
+
+			schemaChain.addFirst(schema_id);
+
+			schema_id = (Long) schemas.get(schema_id).get("base_schema_id");
+		}
+		return schemaChain;
+	}
+
+	private void restoreDerivedSchema(Connection conn, Long schema_id) throws SQLException, IOException, InvalidSchemaError {
+		/* build hashmap of schema_id -> schema properties (as hash) */
+		HashMap<Long, HashMap<String, Object>> schemas = buildSchemaMap(conn);
+
+		/* walk backwards to build linked list with base schema at the
+		 * head, and the rest of the delta schemas following */
+		LinkedList<Long> schemaChain = buildSchemaChain(schemas, schema_id);
```

Comment:
if you prefer to traverse linkedlist, you can just `LinkedHashMap` in the first place, thus you won't bother building chains.